core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1279,9 +1288,11 @@ class Log(@volatile var dir: File,
       val segmentsCopy = logSegments.toBuffer
       // For the earliest and latest, we do not need to return the timestamp.
       if (targetTimestamp == ListOffsetRequest.EARLIEST_TIMESTAMP)
-        return Some(TimestampOffset(RecordBatch.NO_TIMESTAMP, logStartOffset))
+        return Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, logStartOffset,
+          maybeLeaderEpoch(leaderEpochCache.earliestEpoch)))
```

Comment:
The message format may have been changed after logStartOffset. In that case, leaderEpochCache.earliestEpoch may not match logStartOffset. So, we probably want to also make sure the offset corresponding to leaderEpochCache.earliestEpoch matches logStartOffset.